File: ListasDoblementeLigadas/doublyLinkedList.cs
using System;
using System.Collections.Generic;
using System.ComponentModel.Design;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace doublyLinkedList
{
    public class Node
    {
        public Node next, prev;
        public Object item;

        public Node(Object data)
        {
            next = null;
            prev = null;
            item = data;
        }
    }

    public class DoublyLinkedList
    {
        private Node root;

        public bool IsEmpty()
        {
            if (root == null) return true;
            else return false;
        }

        public void InsertAtEnd(Object data)
        {
            if (IsEmpty())
            {
                root = new Node(data);
                return;
            }
            Node reference = root;

            while (reference.next != null) reference = reference.next;

            Node new_node = new Node(data);
            reference.next = new_node;
            new_node.prev = reference;
        }

        public void InsertAtBeginning(Object data)
        {
            if (IsEmpty())
            {
                root = new Node(data);
                return;
            }
            else
            {
                Node new_node = new Node(data);
                root.prev = new_node;
                new_node.next = root;
                root = new_node;
            }
        }

        public void InsertAtPosition(int position, Object data)
        {
            if (position == 0) InsertAtBeginning(data);
            else if (IsEmpty()) Console.WriteLine("Invalid action. The list is empty.");
            else
            {
                Node reference = root;
                for (int i = 0; i < position; i++)
                {
                    if (reference == null)
                    {
                        Console.WriteLine("Invalid action. The list has no element [" + i + "] and forward");
                        return;
                    }
                    reference = reference.next;
                }
                Node new_node = new Node(data);
                if (reference == null) InsertAtEnd(data);
                else
                {
                    reference.prev.next = new_node;
                    new_node.prev = reference.prev;
                    new_node.next = reference;
                    reference.prev = new_node;
                }
            }
        }

        public void DeleteFirstElement()
        {
            if (IsEmpty()) Console.WriteLine("Invalid action. The list is empty");
            else if (root.next == null) root = null;
            else
            {
                root = root.next;
                root.prev = null;
            }
        }

        public void DeleteLastElement()
        {
            if (IsEmpty()) Console.WriteLine("Invalid action. The list is empty");
            else if (root.next == null) root = null;
            else
            {
                Node reference = root;
                while (reference.next != null) reference = reference.next;
                reference.prev.next = null;
            }
        }

        public void DeleteAtPosition(int position)
        {
            if (IsEmpty()) Console.WriteLine("Invlaid action. The list is empty");
            else if (position == 0) DeleteFirstElement();
            else
            {
                Node reference = root;
                for (int i = 0; i < position; i += 1)
                {
                    if (reference.next == null)
                    {
                        Console.WriteLine("Invalid action. The list has no element [" + (i + 1) + "] and forward");
                        return;
                    }
                    reference = reference.next;
                }
                if (reference.next == null) DeleteLastElement();
                else
                {
                    reference.prev.next = reference.next;
                    reference.next.prev = reference.prev;
                }
            }
        }

        public void Search(Object data)
        {
            if (IsEmpty()) Console.WriteLine("Invalid action. The list is empty");
            else
            {
                Node reference = root;
                int count = 0;
                while (true)
                {
                    if (reference == null)
                    {
                        Console.WriteLine("Sorry. The list has no element " + data);
                        return;
                    }
                    else if (data == reference.item)
                    {
                        Console.WriteLine("Element " + data + " located at position [" + count + "]");
                        return;
                    }
                    reference = reference.next;
                    count++;
                }
            }
        }

        public void Show()
        {
            if (IsEmpty()) Console.WriteLine("The list is empty. Insert something");
            else
            {
                Node reference = this.root;
                Console.Write("| ");
                while (reference.next != null)
                {
                    Console.Write(reference.item + " <-> ");
                    reference = reference.next;
                }
                Console.WriteLine(reference.item + " |");
            }
        }

        static void Main()
        {
            DoublyLinkedList list = new DoublyLinkedList();

            /*Put here a menu or the options you would like your list to do.
	        Functions are InsertAtEnd(), InsertAtBeginning(), DeleteFirstElement(), 
            DeleteAtPosition(), Search() and Show()*/

            list.InsertAtEnd("d");
            list.InsertAtEnd("e");
            list.InsertAtEnd("f");
            list.InsertAtEnd("g");
            list.InsertAtBeginning("c");
            list.InsertAtBeginning("b");
            list.InsertAtBeginning("a");
            list.Show();
            list.DeleteAtPosition(7);
            list.InsertAtPosition(7, "a");
            list.DeleteAtPosition(1);
            list.Show();
            list.DeleteAtPosition(5);
            list.Show();
            list.DeleteAtPosition(0);
            list.Show();
            list.InsertAtPosition(1, 1);
            list.DeleteFirstElement();
            list.DeleteLastElement();
            list.Show();
            list.Search(1);

            Console.ReadLine();
        }

    }
}

File: ListasDoblementeLigadas/doublyLinkedList.cpp
#include <iostream>

using namespace std;

class Node{
    public:
        int number;
        const char* word;
        Node *next, *prev;
        Node(int value)
        {
            prev = nullptr;
            next = nullptr;
            number = value; 
            word = nullptr;
        }
        Node(const char* value)
        {
            prev = nullptr;
            next = nullptr;
            number = '\0'; //Number is set to null character to allow
            word = value;  //comparision on conditionals at printing
        }
};

class doublyLinkedList{
    public:
        Node *root;
        doublyLinkedList(){
            root = nullptr;
        }

        bool isEmpty()
        {
            if (this->root == nullptr) return true;
            else return false;
        }

        void insertAtEnd(int data)  //Use polymorphism to help in decision
        {                           //of storing either int or char*
            if(this->isEmpty()){
                root = new Node(data);
                return;
            }
            Node *ref = this->root;
            while (ref->next != nullptr) ref = ref->next;
            Node *new_node = new Node(data);
            ref->next = new_node;
            new_node->prev = ref;
        }
        void insertAtEnd(const char* data)
        {
            if(this->isEmpty()){
                root = new Node(data);
                return;
            }
            Node *ref = this->root;
            while (ref->next != nullptr) ref = ref->next;
            Node *new_node = new Node(data);
            ref->next = new_node;
            new_node->prev = ref;
        }

        void insertAtBeginning(int data)
        {
            if(this->isEmpty()){
                root = new Node(data);
                return;
            }
            else{
                Node *new_node = new Node(data);
                root->prev = new_node;
                new_node->next = root;
                root = new_node;
            }
        }
        void insertAtBeginning(const char* data)
        {
            if(this->isEmpty()){
                root = new Node(data);
                return;
            }
            else{
                Node *new_node = new Node(data);
                root->prev = new_node;
                new_node->next = root;
                root = new_node;
            }
        }

        void insertAtPosition(int position, int data)
        {
            if (position == 0) this ->insertAtBeginning(data);
            else if (this->isEmpty()) cout << "Invalid action. The list is empty\n";
            else{
                Node *ref = root;
                for(int i=0; i<position; i++){
                    if(ref == nullptr){
                        cout << "Invalid action. The list has no element ["<<position<<"] and forward\n";
                        return;
                    }
                    ref = ref->next;
                }
                Node *new_node = new Node(data);
                if(ref == nullptr) this->insertAtEnd(data);
                else{
                    ref->prev->next = new_node;
                    new_node->prev = ref->prev;
                    new_node->next = ref;
                    ref->prev = new_node;
                }
            }
        }

        void insertAtPosition(int position, const char* data)
        {
            if (position == 0) this ->insertAtBeginning(data);
            else if (this->isEmpty()) cout << "Invalid action. The list is empty\n";
            else{
                Node *ref = root;
                for(int i=0; i<position; i++){
                    if(ref == nullptr){
                        cout << "Invalid action. The list has no element ["<<position<<"] and forward\n";
                        return;
                    }
                    ref = ref->next;
                }
                Node *new_node = new Node(data);
                if(ref == nullptr) this->insertAtEnd(data);
                else{
                    ref->prev->next = new_node;
                    new_node->prev = ref->prev;
                    new_node->next = ref;
                    ref->prev = new_node;
                }
            }
        }

        void deleteFirstElement()
        {
            if(this->isEmpty()) cout << "Invalid action. The list is empty\n";
            else if (root->next == nullptr) root = nullptr;
            else{
                root = root->next;
                root->prev = nullptr;
            }
        }

        void deleteLastElement()
        {
            if(this->isEmpty()) cout << "Invalid action. The list is empty\n";
            else if (root->next == nullptr) root = nullptr;
            else{
                Node *ref = root;
                while(ref->next != nullptr) ref = ref ->next;
                ref->prev->next = nullptr;
            }
        }

        void deleteAtPosition(int position){
            if (this->isEmpty()) cout << "Invalid action. The list is empty\n";
            else if(position == 0) this ->deleteFirstElement();
            else{
                Node *ref = root;
                for (int i = 0; i < position; i++)
                {
                    if(ref->next == nullptr){
                        cout << "Invalid action. The list has no element ["<<position<<"] and forward\n";
                        return;
                    }
                    ref = ref->next;
                }
                if(ref->next == nullptr) this->deleteLastElement();
                else{
                    ref->prev->next = ref->next;
                    ref->next->prev = ref->prev;
                }
            }
        }

        void search(int data)
        {
            if(this->isEmpty()) cout << "Invalid action. The list is empty\n";
            else{
                Node *ref = root;
                int count = 0;
                while (true){
                    if(ref == nullptr){
                        cout << "Sorry. The list has no element "<< data <<endl;
                        return;
                    }
                    else if (data == ref->number && ref->number!='\0')
                    {
                        cout << "Element "<< data << " located at position [" << count <<"]"<<endl;
                        return;
                    }
                    ref = ref->next;
                    count++;
                }
            }
        }
        void search(const char* data)
        {
            if(this->isEmpty()) cout << "Invalid action. The list is empty\n";
            else{
                Node *ref = root;
                int count = 0;
                while (true){
                    if(ref == nullptr){
                        cout << "Sorry. The list has no element "<< data <<endl;
                        return;
                    }
                    else if (data == ref->word)
                    {
                        cout << "Element "<< data << " located at position [" << count <<"]"<<endl;
                        return;
                    }
                    ref = ref->next;
                    count++;
                }
            }
        }

        void show()
        {
            if(this->isEmpty()) 
                cout<<"The list is empty. Insert something :)\n";
            else{
                Node *ref = this->root;
                cout << "| ";
                while (ref->next != nullptr)
                {
                    if(ref->number!='\0') cout << ref->number << " <-> ";
                    else cout << ref->word << " <-> ";
                    ref = ref->next; 
                }
                if(ref->number!='\0') cout << ref->number << " |\n";
                else cout << ref->word << " |\n";
            }
        }
};

int main(){
    doublyLinkedList *list = new doublyLinkedList();

    //Put here a menu or the options you would like your list to do.

    list->insertAtEnd("d");
    list->insertAtEnd("e");
    list->insertAtEnd("f");
    list->insertAtEnd("g");
    list->insertAtBeginning("c");
    list->insertAtBeginning("b");
    list->insertAtBeginning("a");
    list->show();
    list->deleteAtPosition(7);
    list->deleteAtPosition(1);
    list->show();
    list->deleteAtPosition(5);
    list->show();
    list->deleteAtPosition(0);
    list->show();
    list->insertAtPosition(1,1);
    list->show();
    list->search(0);
    
    system("pause");
    return 0;
}